### PR TITLE
[PM-2733] Master Password Does not Reprompt onEdit after Cancelling M…

### PR DESCRIPTION
**Pull request by cagonzalezcs**
## Type of change

```
- [x] Bug fix
- [ ] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

During testing of #5635 , QA found an issue where copying the password of a vault item, that requires a master password re-prompt, would allow a user to edit the item without a re-prompt for the edit action. In review of the bug, I found that we were storing the most recent result from a master password re-prompt. 

If a user enacts a copy of a vault item password, we store the result of the re-prompt that allows the user to finish the copy action. Since this copy action does not unload the `view` component used to display the vault item, the value indicating that a re-prompt has occurred never gets set back to false. 

I've put together some changes that address this issue. First, anytime we are triggering a master password re-prompt on an action that will not unload the `view` component, we now disregard the result of the re-prompt, forcing a subsequent re-prompt on sensitive actions. The actions that trigger a re-prompt but do not refresh the component include:

- Toggling visibility of a vault item password
- Toggling visibility of a vault item credit card number
- Toggling visibility of a vault item credit card security code
- Copying a value 

I've also done some refining of the user experience, some details that I'd like to ask @danielleflinn to review and verify are desired from a UX perspective. These can be removed or altered if necessary. The following adjustments have been put in place:

- When toggling a protected field (password, cc number, cvv), if the field is currently not visible then we will re-prompt users for the password (if the item has re-prompt active). If the field is visible, and the user clicks toggle we will not re-prompt the user and set the visibility for the field to hide the element.
- When attempting to copy a protected field, if the field is not currently visible we will re-prompt the user in order to allow the copy. If the field has been toggled to a visible state, then we will not re-prompt the user if they opt to copy the value using the copy button.

The behavior described above can be see in the video attached to this PR.

## Code changes

- **libs/angular/src/vault/components/view.component.ts:** Modified the `promptPassword` method to indicate whether the `passwordReprompted` value should be stored. Will ignore the return value of the password re-prompt if indicated to not store the value. Added a method `copiedFieldIsVisible` which triggers when `copy` is run. This ensures that a password re-prompt only occurs if the field is currently not visible to the user. Added a method `failedPasswordRepromptForFieldToggle` which is used when toggling the visibility of a vault item password, credit card number, or credit card security code. This ensures that users are required to pass a re-prompt if the field is hidden, but do not have to re-prompt in order to hide the field from a visible state.
- **libs/angular/src/vault/services/password-reprompt.service.ts:** Modified the method `protectedFields` to reference the newly added enum `CipherRepromptProtectedFields` as an array of values.
- **libs/common/src/vault/enums/cipher-reprompt-protected-fields.enum.ts:** Added an enum for `CipherRepromptProtectedFields` which allows us to share these values between the `PasswordRepromptService` and `ViewComponent` classes.

### Original Behavior

https://github.com/bitwarden/clients/assets/16629865/468e7b09-a3f2-437a-b2d0-cd69eface122

### Fixed Behavior

https://github.com/bitwarden/clients/assets/16629865/4690bee1-cf2a-4566-9a41-ee48c08f506b

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

**Comment by danielleflinn** [comment]
The updated user experience looks good. Thanks for working on this one!

File: libs/common/src/enums/password-reprompt-protected-fields.enum.ts
Comment (djsmith85) [inline libs/common/src/enums/password-reprompt-protected-fields.enum.ts:1]: Please move this file into `libs/common/src/vault/enums/` and possibly rename it to `cipher-reprompt-protected-fields.enum.ts` , as we might accept other ways to pass the re-prompt

**Review by djsmith85 — CHANGES_REQUESTED** [review]

File: libs/common/src/enums/password-reprompt-protected-fields.enum.ts
Comment (cagonzalezcs) [reply libs/common/src/enums/password-reprompt-protected-fields.enum.ts:1]: @djsmith85 

You've got it, this change has been pushed up.

**Comment by cagonzalezcs** [comment]
@djsmith85 

When you find some time, is there anything else I should improve within this implementation?

File: libs/angular/src/vault/components/view.component.ts [hunk 1]
```diff
@@ -281,6 +282,10 @@ export class ViewComponent implements OnDestroy, OnInit {
     }
   }
 
+  async failedPasswordRepromptForFieldToggle(isFieldShowing = false): Promise<boolean> {
```
Comment (djsmith85) [inline libs/angular/src/vault/components/view.component.ts:285]: Just from the methodName it is not clear what this method does and what has or is failing.

File: libs/angular/src/vault/components/view.component.ts [hunk 2]
```diff
@@ -347,6 +353,20 @@ export class ViewComponent implements OnDestroy, OnInit {
     return true;
   }
 
+  copiedFieldIsVisible(fieldType: string): boolean {
```
Comment (djsmith85) [inline libs/angular/src/vault/components/view.component.ts:356]: Maybe just my personal preference, but I think `IsCopiedFieldVisible` is neater.

**Review by djsmith85 — COMMENTED** [review]
Just two small naming-related things. 

Through the file-move I initially requested, the code-ownership is now with team-vault-dev, so I'm going to let them do the final review.

File: libs/angular/src/vault/components/view.component.ts (hunk 2 above)
Comment (cagonzalezcs) [reply libs/angular/src/vault/components/view.component.ts]: Sounds good, I've modified the name of this method to `isCopiedFieldVisible` as suggested.

**Review by djsmith85 — APPROVED** [review]
@cagonzalezcs Thank you for the name changes. This clarifies what the methods are doing.

Leaving the final review to `team-vault-dev`

**Comment by cagonzalezcs** [comment]
Closing this PR as per discussion within PM-2733's ticket. We're approaching the master password re-prompt process in a different manner down the line, and incorporation of this code will force us to rework certain aspects of the affected files later one.

**Pull request closed without merging**